### ISSUE TITLE
Cannot correctly detect doc if using single quote in Inferior Python mode.

### ISSUE THREAD

**Issue by yanghaoxie** [davidshepherd7/electric-operator#87]
If your issue is a bug, please provide me with some information to help reproduce it. If it's a feature request then
you can ignore this template.

## Reproducing

-  scenario 1.
If I first input `"`, then input `/`, it will produce `>>> "/`

-  scenario 2.
If I first input `'`, then input `/`, it will produce `>>> ' / `

You see, in scenario 1, there is no extra space, since I trying insert string. However in scenario 2, I use single quote, there is an extra space.

## Other useful info

* Emacs version: 27.0.50
* electric-operator version: 1.0.0
* the value of `electric-operator-enable-in-docs` is `nil`

**Comment by davidshepherd7** [comment]
Hi, thanks for reporting this and for the reproduction details :)

Unfortunately I think this is due to a bug in inferior python mode's parsing. electric-operator relies on the major mode to tell it if it is currently inside a string or not, and for some reason inferior python mode claims that `>>> '` is not inside a string. I don't think there's really much I can do about it without implementing a whole python parser :disappointed:.

**Comment by yanghaoxie** [comment]
Thank for you for explain. I guess I have to file a issue to Emacs develop group.

**Comment by yanghaoxie** [comment]
Hi! 
I wonder is there a simple workaround that we test `'` ourself (in `electric-operator-in-docs?`) instead using `syntax-ppss` in `inferior python mode `.

Maybe we can just add the test like this?

```
(defun electric-operator-in-docs? ()
  "Check if we are inside a string or comment"
  (or (nth 8 (syntax-ppss))
      (my-own-single-quote-test-in-inferior-python-mode)))
```

**Comment by davidshepherd7** [comment]
Yeah something like that is what I meant, the issue is that in order to
correctly implement my-own-single-quote-test you would need to write a
parser. Here are some examples to think about:

```
>>> ' '

>>> " ' "

>>> '\'

>>> ">>> ' "

>>> """
' '

>>> " ' " '
```

In fact it's even harder than I thought because if the buffer contains

```
>>> x = '
Syntax error blah blah blah
>>> [point here]
```

then we need to recognise that the unbalanced quote doesn't make the rest
of the buffer into a string.

If you're still interested you could give it a try and send a PR (although
make sure it has tests for all of those cases and all similar ones you can
come up with), but I think it would be much easier to fix the existing
parser. Maybe it's even already fixed in one of the unstable releases of
emacs.

**Comment by yanghaoxie** [comment]
Thank you so much for you comprehensive explanation. Sounds like it is much harder than I thought.

By the way, I am using the Emacs 27.0.50, it seems like that this is still a issue.

Anyway, thank you very much! :+1:

**Comment by davidshepherd7** [comment]
No problem, glad it helped :slightly_smiling_face: 

> By the way, I am using the Emacs 27.0.50, it seems like that this is still a issue.

Oh, that's a shame.
